File: lesson-3_dynamic-programming/0-1-knapsack-problem/recurrence/main.go
package main

import (
    "fmt"
    "math"
)

func main() {
    var n, balance int
    fmt.Scan(&n, &balance)

    values := make([]int, n)
    like := make([]int, n)
    fav := make([]int, balance+1)
    for i := 0; i < n; i++ {
        fmt.Scan(&values[i], &like[i])
    }
    for i := 1; i <= balance; i++ {
        if values[0] <= i {
            fav[i] = like[0]
        }
    }
    for i := 1; i < n; i++ {
        for j := balance; j >= 1; j-- {
            if values[i] <= j {
                fav[j] = int(math.Max(float64(fav[j]), float64(fav[j-values[i]]+like[i])))
            }
        }
    }
    fmt.Println(fav[balance])
}

File: lesson-3_dynamic-programming/0-1-knapsack-problem/recurrence/main.cpp
#include <iostream>
#include <vector>
#include <fmt/core.h>

int main() {
    int n, balance;
    std::cin >> n >> balance;

    auto values = std::vector<int>(n);
    auto like = std::vector<int>(n);
    auto fav = std::vector<int>(balance + 1);
    for (int i = 0; i < n; i++) {
        std::cin >> values[i] >> like[i];
    }
    for (int i = 1; i <= balance; i++) {
        if (values[0] <= i) {
            fav[i] = like[0];
        }
    }
    for (int i = 1; i < n; i++) {
        for (int j = balance; j >= 1; j--) {
            if (values[i] <= j) {
                fav[j] = std::max(fav[j], fav[j - values[i]] + like[i]);
            }
        }
    }
    fmt::print("{}\n", fav[balance]);
    return 0;
}
